deepsubtree_test.go

```diff
@@ -125,3 +125,106 @@ func TestDeepSubtreeWithUpdates(t *testing.T) {
+// correct roots
+func TestDeepSubtreeWWithAddsAndDeletes(t *testing.T) {
+	require := require.New(t)
+	getTree := func() *MutableTree {
+		tree, err := getTestTree(5)
+		require.NoError(err)
+
+		tree.Set([]byte("b"), []byte{2})
+		tree.Set([]byte("a"), []byte{1})
+
+		_, _, err = tree.SaveVersion()
+		require.NoError(err)
+		return tree
+	}
+	tree := getTree()
+	fmt.Println("PRINT TREE")
+	_ = printNode(tree.ndb, tree.root, 0)
+	fmt.Println("PRINT TREE END")
+
+	subsetKeys := [][]byte{
+		[]byte("b"),
+	}
+	rootHash := tree.root.hash
+	mutableTree, err := NewMutableTree(db.NewMemDB(), 100, false)
+	require.NoError(err)
+	dst := DeepSubTree{mutableTree}
+	for _, subsetKey := range subsetKeys {
+		ics23proof, err := tree.GetMembershipProof(subsetKey)
+		require.NoError(err)
+		err = dst.AddExistenceProof(ics23proof.GetExist())
+		require.NoError(err)
+	}
+
+	keysToAdd := [][]byte{
+		[]byte("c"), []byte("d"),
+	}
+	valuesToAdd := [][]byte{
+		{3}, {4},
+	}
+	// Add non-existence proofs for keys we expect to add later
+	for _, keyToAdd := range keysToAdd {
+		ics23proof, err := tree.GetNonMembershipProof(keyToAdd)
+		require.NoError(err)
+		dst_nonExistenceProof, err := convertToDSTNonExistenceProof(tree, ics23proof.GetNonexist())
+		require.NoError(err)
+		dst.AddNonExistenceProof(dst_nonExistenceProof)
+		require.NoError(err)
+		dst.BuildTree(rootHash)
+		require.NoError(err)
+	}
+	dst.SaveVersion()
+
+	fmt.Println("PRINT DST TREE")
+	_ = dst.printNodeDeepSubtree(dst.ImmutableTree.root, 0)
+	fmt.Println("PRINT DST TREE END")
+
+	// Check root hashes are equal
+	require.Equal(dst.root.hash, tree.root.hash)
+
+	require.Equal(len(keysToAdd), len(valuesToAdd))
+	// Add all the keys we intend to add and check root hashes stay equal
+	for i := range keysToAdd {
+		keyToAdd := keysToAdd[i]
+		valueToAdd := valuesToAdd[i]
+		dst.Set(keyToAdd, valueToAdd)
+		dst.SaveVersion()
+		err = dst.BuildTree(dst.root.hash)
+		require.NoError(err)
+		fmt.Println("PRINT DST TREE")
+		_ = dst.printNodeDeepSubtree(dst.ImmutableTree.root, 0)
+		fmt.Println("PRINT DST TREE END")
+		tree.Set(keyToAdd, valueToAdd)
+		tree.SaveVersion()
+		fmt.Println("PRINT TREE")
+		_ = printNode(tree.ndb, tree.root, 0)
+		fmt.Println("PRINT TREE END")
+
+		// Check root hashes are equal
+		require.Equal(dst.root.hash, tree.root.hash)
+	}
+
+	// Delete all the keys we added and check root hashes stay equal
+	for i := range keysToAdd {
+		keyToAdd := keysToAdd[i]
+		dst.Remove(keyToAdd)
+		dst.SaveVersion()
+		err = dst.BuildTree(dst.root.hash)
+		require.NoError(err)
+		fmt.Println("PRINT DST TREE")
+		_ = dst.printNodeDeepSubtree(dst.ImmutableTree.root, 0)
+		fmt.Println("PRINT DST TREE END")
+		tree.Remove(keyToAdd)
+		tree.SaveVersion()
+		fmt.Println("PRINT TREE")
+		_ = printNode(tree.ndb, tree.root, 0)
+		fmt.Println("PRINT TREE END")
+
+		// Check root hashes are equal
+		require.Equal(dst.root.hash, tree.root.hash)
+	}
+}
```

Comment:
As @musalbas mentioned add a bunch of (potentially larger) randomized tests, that:
- batch-add
- batch-remove

IMO, looking into the SMT tests for inspiration (even if it does not do rebalancing), is a good idea. Add some light fuzzing maybe.

Reply:
Also, add table-driven tests for each public method.

Reply:
Documented in #15